coreneuron/io/lfp.hpp

```diff
@@ -0,0 +1,226 @@
+}
+
+template <typename Point3D, typename F>
+F norm(const Point3D& p1) {
+    return std::sqrt(dot<Point3D, F>(p1, p1));
+}
+
+template <typename Point3D, typename F>
+Array3<F> barycenter(const Point3D& p1, const Point3D& p2) {
+    return {0.5 * (p1[0] + p2[0]), 0.5 * (p1[1] + p2[1]), 0.5 * (p1[2] + p2[2])};
+}
+
+template <typename Point3D, typename F>
+Array3<F> axp(const Point3D& p1, const F alpha, const Point3D& p2) {
+    return {p1[0] + alpha * p2[0], p1[1] + alpha * p2[1], p1[2] + alpha * p2[2]};
+}
+
+/**
+ *
+ * \tparam Point3D Type of 3D point
+ * \tparam F Floating point type
+ * \param e_pos electrode position
+ * \param seg_pos segment position
+ * \param radius segment radius
+ * \param f conductivity factor 1/([4 pi] * [conductivity])
+ * \return Resistance of the medium from the segment to the electrode.
+ */
+template <typename Point3D, typename F>
+F point_source_lfp_factor(const Point3D& e_pos, const Point3D& seg_pos, const F radius, const F f) {
+    nrn_assert(radius >= 0.0);
+    Array3<F> es = axp(e_pos, -1.0, seg_pos);
+    return f / std::max(norm<Point3D, F>(es), radius);
+}
+
+/**
+ *
+ * \tparam Point3D Type of 3D point
+ * \tparam F Floating point type
+ * \param e_pos electrode position
+ * \param seg_pos segment position
+ * \param radius segment radius
+ * \param f conductivity factor 1/([4 pi] * [conductivity])
+ * \return Resistance of the medium from the segment to the electrode.
+ */
+template <typename Point3D, typename F>
+F line_source_lfp_factor(const Point3D& e_pos,
+                         const Point3D& seg_0,
+                         const Point3D& seg_1,
+                         const F radius,
+                         const F f) {
+    nrn_assert(radius >= 0.0);
+    Array3<F> dx = axp(seg_1, -1.0, seg_0);
+    Array3<F> de = axp(e_pos, -1.0, seg_0);
+    F dx2(dot<Array3<F>, F>(dx, dx));
+    F dxn(std::sqrt(dx2));
+    if (dxn < 1.0e-20) {
+        return point_source_lfp_factor(e_pos, seg_0, radius, f);
+    }
+    F de2(dot<Array3<F>, F>(de, de));
+    F mu(dot<Array3<F>, F>(dx, de) / dx2);
+    Array3<F> de_star(axp(de, -mu, dx));
+    F de_star2(dot<Array3<F>, F>(de_star, de_star));
+    F q2(de_star2 / dx2);
+
+    F delta(mu * mu - (de2 - radius * radius) / dx2);
+    F one_m_mu(1.0 - mu);
+    auto log_integral = [&q2, &dxn](F a, F b) {
+        if (std::abs(q2) < std::numeric_limits<F>::epsilon()) {
+            if (a * b <= 0) {
+                throw std::invalid_argument("Log integral: invalid arguments " + std::to_string(b) +
+                                            " " + std::to_string(a) +
+                                            ". Likely electrode exactly on the segment and "
+                                            "no flooring is present.");
+            }
+            return std::abs(std::log(b / a)) / dxn;
+        } else {
+            return std::log((b + std::sqrt(b * b + q2)) / (a + std::sqrt(a * a + q2))) / dxn;
```

Comment:
I will not check this since I already checked last time... except if you changed something here

Reply:
There has not been material changes since your review I believe.